15 show book names instead of isbn load new csv

Magener opened the pull request.

Magener committed:
feat(Presenter.py, GraphSynthesizer.py):
Added automated presentation of book name if it is available, otherwise present ISBN in graph.

Magener committed:
feat(CSVSensor.py, book_ddl.sql):
Changed RatingSensor to be a generic CSV Sensor - for reuse as a book sensor.
Renamed book_ddl.sql

Magener committed:
feat(review_ddl.sql):
Renamed review_ddl.sql

Magener committed:
feat(JSONReceiver.py, MessageHandlingStrategy.py, RatingReceiver.py): basic refactoring.

Magener committed:
feat(Book.py, BookReceiver.py, MessageValidation.py):
Basic refactoring

Magener committed:
feat(RatingReceiver.py,Review.py):
fixed RatingReceiver.py to work.

Magener committed:
feat(AsyncPostgresConnection.py, BatchProcessor.py, BookOperations.py, consts.py, log.py):
Basic batch processor that doesn't have processing mechanisms

Elon-Abulafia commented on book_ddl.sql:
```diff
@@ -0,0 +1,12 @@
+CREATE TABLE hafifa.unprocessed_books (
+    isbn VARCHAR(20) PRIMARY KEY,
+    book_title VARCHAR(255),
+    book_author VARCHAR(255),
+    year_of_publication INTEGER,
+    publisher VARCHAR(255)
+);
+
+CREATE TABLE hafifa.book (
+    book_id SERIAL PRIMARY KEY,
+    book_name VARCHAR(255)
+);
```
Maybe you should try using `sqlalchemy` instead of writing all of the queries and such like this. (Also in the code you write queries using strings, which is okay, but consider it)

Magener replied:
I agree and have decided to shift to sql alchemy.
it is ORM, but sync. meaning there would be some additional work
in making it async. postponed for now for that reason.

Elon-Abulafia commented on presenter/GraphSynthesizer.py:
```diff
@@ -0,0 +1,8 @@
+    return row["book_id"]
+
+
+def compute_presented_name(book_dataframe):
+    return book_dataframe.apply(get_book_identifier, axis=1)
```
If your function is only one line, and is used only in a single place, maybe it shouldn't be a function..

Magener replied:
I plan on inlining get_book_identifier, other than that, its parent is already too big (you've already c ommented about it).
I think that after refactoring parent, it will be clearer wether or not this function should be extracted or not.

Elon-Abulafia commented on presenter/GraphSynthesizer.py:
```diff
@@ -0,0 +1,8 @@
+def get_book_identifier(row) -> str:
+    if "book_name" in row:
+        return row["book_name"]
+    return row["book_id"]
```
Again this seems very unintuitive, try to do better than that and consider if you even need a function.

Magener replied:
I have decided to inline it to the parent, possibly into a 1 liner, I have to see how readable it is in order to decide wether it should be a 1 liner or not.

Elon-Abulafia commented on presenter/Presenter.py:
```diff
@@ -2,20 +2,21 @@
 plt.tight_layout()
-plt.show()
\ No newline at end of file
+plt.show()
```
In general, don't write code in the outer scope. Always put code inside a function or a scope of some sort.

Magener replied:
I agree, will be extracted now.

Elon-Abulafia commented on receiver/MessageHandlingStrategy.py:
```diff
@@ -0,0 +1,6 @@
+from typing import Protocol
+
+
+class MessageHandlingStrategy(Protocol):
+    async def on_message(self, data: dict) -> None:
+        pass
```
Not sure if the `Protocol` class is necessary if you only have a single function. Maybe `ABC` is more fitting here?

Magener replied:
The difference really is negligible.
	However, ABC does enforce its interface which is desired
	in this case.
	Protocol is when inheritance is bad, but there is value in defining the required structure.
	In this case since structure is strictly defined & inheritance would make sense,
		you are right, ABC is the better choice.

Elon-Abulafia commented on receiver/MessageValidation.py:
```diff
@@ -6,10 +6,24 @@ def validate_message(msg) -> None:
     except ValueError as e:
-        raise ValueError(f"Value provided was not parsable! {rating_data} {e}")
\ No newline at end of file
+        raise ValueError(f"Value provided was not parsable! {data} {e}")
```
Very nice way yo handle multiple data types 👍

Elon-Abulafia commented on receiver/book/Book.py:
```diff
@@ -0,0 +1,19 @@
+
+
+async def commit_raw_book(isbn, book_title, book_author, year_of_publication, publisher):
+    connection = await AsyncPostgresConnection().get_connection()
+
+    QUERY = "INSERT INTO hafifa.unprocessed_books" \
+            "(isbn, book_title, book_author, year_of_publication, publisher)" \
+            " VALUES ($1, $2, $3, $4, $5);"
+
+    result_message = await connection.execute(QUERY, isbn, book_title, book_author, year_of_publication, publisher)
+
+    insert_successful = result_message.startswith("INSERT")
+
+    if insert_successful:
+        logger.debug(f"Regstered {(isbn, book_title, book_author, year_of_publication, publisher)} in database.")
+    else:
+        raise RuntimeError(f"Insertion has failed: {result_message}")
```
Pretty sure if you move the entire DB interaction logic into a different class this will be simplified significantly.

Magener replied:
would change after moving to SQL alchemy therefore postponed.

Elon-Abulafia commented on receiver/JSONReceiver.py:
```diff
@@ -0,0 +1,37 @@
+    def start(self) -> None:
+        asyncio.run(self.__consume_json_messages())
+
+    async def __consume_json_messages(self) -> None:
```
```Suggestion
async def _consume_json_messages(self) -> None:
```
Private functions in python start with a single underscore.

Magener replied:
I remember protected being a single underscore and private being 2 underscores.

Elon-Abulafia commented on receiver/JSONReceiver.py:
```diff
@@ -0,0 +1,37 @@
+
+        await consumer.start()
+
+        try:
+            async for message in consumer:
+                data = json.loads(message.value)
+                logger.info(f"Received Data: {data}")
+                await self.handler_strategy.on_message(data)
+        finally:
+            await consumer.stop()
```
Why stop the consumer if an error occurred?
Maybe try to handle exceptions here?

Magener replied:
You are right, this was planned to be dealt with in this branch, and it will.

Elon-Abulafia commented on receiver/JSONReceiver.py:
```diff
@@ -0,0 +1,37 @@
+import asyncio
+import json
+
+from aiokafka import AIOKafkaConsumer
+
+from receiver.MessageHandlingStrategy import MessageHandlingStrategy
+from receiver.consts import RATING_TOPIC_NAME, KAFKA_BROKER_URL
+from receiver.log import logger
```
Arrange imports nicely.

Magener replied:
Not a high priority right now, will find time for it later.

Elon-Abulafia commented on receiver/JSONReceiver.py:
Overall very good job.

Magener committed:
feat(book_ddl.sql):
changed ddl configuration slightly.

Magener committed:
feat(BatchProcessor.py, BookOperations.py, consts.py, ExitCatcher.py):
added batch processing of  book data at regular intervals

Magener committed:
feat(Review.py, GraphSynthesizer.py):
Adjusted query to send book join book names to query if possible. Adjusted presenter to accomodate for a case in which book name is unavailable.

Magener committed:
feat(book_ddl.sql, Review.py, review_ddl.sql):
some refactoring & moving files.

Magener committed:
feat(...): pull changes from parent branch & merge conflicts.

Magener committed:
feat(JSONReceiver.py, RatingReceiver.py, Review.py):
fixed basic errors.

Magener committed:
feat(AsyncPostgresConnection.py, BatchProcessor.py, BookOperations.py, GraphSynthesizer.py, MessageHandlingStrategy.py, Presenter.py): some refactoring.

Magener committed:
feat(...): added error handling capabilities for queries. added folders to create_requirements_files.bat, and some folder extraction for better readability.

Magener committed:
feat(requirements.txt):
added additional requirements for each folder.

Magener committed:
feat(__init__.py):
added __init__.py files - to allow for microservices to be containerized.

Magener committed:
feat(AsyncPostgresConnection.py, Book.py, DBBase.py, Rating.py, Review.py, review_ddl.sql):
Basic migration of InsightAPI to sqlalchemy.

Magener committed:
feat(./kafka):
moved to a different file to do impact on importation.

Magener committed:
feat(AsyncPostgresConnection.py, BookDBInteractions.py, BookOperations.py, BookReceiver.py, UnprocessedBook.py): completed migration to sqlalchemy.

Magener committed:
feat(BookDBInteractions.py, BookModel.py, BookOperations.py, Presenter.py, RatingDBInteractions.py, RatingModel.py, Review.py, UnprocessedBookModel.py):
renamed all models to contain a Model keyword for clarity.

Magener committed:
feat(BookDBInteractions.py, BookReceiver.py, RatingDBInteractions.py, RatingModel.py, RatingReceiver.py, Review.py, UnprocessedBookModel.py):
Refactored redundent models and provided a simpler syntax.

Magener committed:
feat(RatingModel.py):
added basic relationship

Magener committed:
feat(...):
moved shared files to a separate folder.
prepared extraction of DB repository to a separate microservice.

Magener committed:
feat(.env, .gitignore): removed .env & updated file.

Magener committed:
feat(RatingModel.py,RatingReceiver.py):
Extracted transformation from sql_alchemy score, for future separation to a microservice.

Magener committed:
feat(...):
some extraction to folders.
Updated requirements.txt as necessary.

Magener committed:
feat(requirements.txt):
added requirements.txt

Magener committed:
feat(requirements.txt):
added requirements.txt

Magener committed:
feat(InitializePackages.py):
a basic script for adding __init__.py files.

Magener committed:
feat(__init__.py):
added __init__.py file.

Magener committed:
feat(ddl.sql):
merged ddl to a single file.

Magener committed:
feat(.gitignore):
updated .gitignore.

Magener committed:
feat(...):
renamed folders to match microservice names.

Magener committed:
feat(BatchProcessor.py):
fixed wrong async behavior.

Magener committed:
feat(BatchProcessor.py, BookRepository.py, BookTransformer.py, CommandParser.py, DBEndpoint.py, InsightAPI.py, RatingRepository.py, RatingTransformer.py):
Added basic command parsing, for future separation of sql to a micro service.
